### PR TITLE
Fix DataGrid native aot crash when sorting.

### LINKED ISSUE #14059: Click [decimal/int32/enum] DataGrid Header crash when publish aot.
## Describe the bug

When I click the datagrid column header to sort which display decimal/enum/int32 field, the application will crash. 
It works normal when not aot mode.
It works normal when dotnet run with `PublishAot=true`
It works normal when column data type is `string`/`double`

## To Reproduce
I create a simple demo to reproduce this bug:
1. create a avalonia data grid demo.
2. create a `TestModel` which has decimal/int32/enum fields.
3. create a DataGrid to display TestModel data.
4. add `<PublishAot>true</PublishAot>` to csproj.
5. publish the application with NativeAot:`dotnet publish -c Release`
6. execute the published application:`.\bin\Release\net8.0\win-x64\publish\AvaloniaDataGrid.exe`
7. Click header to sort.

I have create a repo to reproduce that. https://github.com/dameng324/AvaloniaDataGridDemo

## Expected behavior
do not crash when

## Screenshots
you can simple download the code and reproduce it.

## Environment 

- OS: Windows 
- Avalonia-Version: 11.0.6
- dotnet sdk version: 8.0.100

## Additional context
I have tried many types. result:
|type| dotnet publish aot and run manually|
|---|---------------------------------|
|int32| crash|
|decimal|crash|
|enum|crash|
|double|normal|
|string|normal|

### PR DESCRIPTION
**Pull request by dameng324**
## What does the pull request do?
When sorting datagrid with a non string type field, the application will crash as no metadata of `Comparer<int>`. This PR fix it.

Fixed: https://github.com/AvaloniaUI/Avalonia/issues/14059

## What is the current behavior?
- When user does not provide the ViewModel meta data (by [DynamicDependcy] or Rd.xml etc), the application does not crash when sorting int field. but the sort function is broken.

- When user provide the ViewModel meta data (by [DynamicDependcy] or Rd.xml etc), the application will crash when sorting int field.

## What is the updated/expected behavior with this PR?
- When user does not provide the ViewModel meta data (by [DynamicDependcy] or Rd.xml etc), the application will behave same as before when sorting.

- When user provide the ViewModel meta data (by [DynamicDependcy] or Rd.xml etc), the application will work normal when sorting `int`/`int?`/`Enum`/`Enum?` and other type which implements `IComparable` interface when using aot. Other type which does not implements `IComparable` should not crash, only does nothing when sorting.

## How was the solution implemented (if it's not obvious)?

Old solution:
```cs
return (typeof(Comparer<>).MakeGenericType(type).GetProperty("Default")).GetValue(null, null) as IComparer;
```

My solution:
```cs
#if NET6_0_OR_GREATER
if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>) && type.GetGenericArguments()[0].IsAssignableTo(typeof(IComparable)))
    return Comparer<object>.Create((x, y) => 
    {
        if (x == null)
            return y == null ? 0 : -1;
        else
            return (x as IComparable)!.CompareTo(y);
    });
else if (type.IsAssignableTo(typeof(IComparable)))
    return Comparer<object>.Create((x, y) => (x as IComparable)!.CompareTo(y));
else
    return Comparer<object>.Create((x, y) => 0);
#else
    return (typeof(Comparer<>).MakeGenericType(type).GetProperty("Default")).GetValue(null, null) as IComparer;
#endif
```

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Fixed issues
Fixed: https://github.com/AvaloniaUI/Avalonia/issues/14059

### CODE REVIEW

**Comment by dameng324** [comment]
@cla-avalonia agree

File: src/Avalonia.Controls.DataGrid/Collections/DataGridSortDescription.cs [hunk 1]
```diff
@@ -170,8 +170,29 @@ private IComparer GetComparerForType(Type type)
...
+                    {
+                        if (x == null)
+                            return y == null ? 0 : -1;
+                        else
+                            return (x as IComparable)!.CompareTo(y);
+                    });
+                else if (type.IsAssignableTo(typeof(IComparable)))
+                    return Comparer<object>.Create((x, y) => (x as IComparable)!.CompareTo(y));
+                else
+                    return Comparer<object>.Create((x, y) => 0); //avoid using reflection to avoid crash on AOT
```
Comment (maxkatz6) [inline src/Avalonia.Controls.DataGrid/Collections/DataGridSortDescription.cs:190]: On NET6_0_OR_GREATER you can check for this property whether reflection is supported: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimefeature.isdynamiccodesupported?view=net-8.0

On AOT builds IsDynamicCodeSupported returns false, and this fallback would make sense there.

Comment (maxkatz6) [reply src/Avalonia.Controls.DataGrid/Collections/DataGridSortDescription.cs:190]: This way we also won't regress any application that realizes on this reflection sorting right now.

Comment (dameng324) [reply src/Avalonia.Controls.DataGrid/Collections/DataGridSortDescription.cs]: Nice suggestion. Done.

**Comment by dameng324** [comment]
User can not use sort feature when they does not provide metadata of the type when using aot, Can we do better to avoid user make this kind of mistake? such as add some aot compile warning, or add [DynamicallyAccessedMembersAttribute] or [DynamicDependency] to somewhere? @maxkatz6 do you have any idea on it?

**Review by maxkatz6 — APPROVED** [review]

**Comment by maxkatz6** [comment]
@dameng324 we could generate `DynamicDependency` based on `DataGridColumn.SortMemberPath` type. See https://github.com/AvaloniaUI/Avalonia/issues/14085. As an extra attribute on https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls.DataGrid/DataGridColumn.cs#L1124C23-L1124C37

Ideally, AOT apps would utilize `CustomSortComparer` instead https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls.DataGrid/DataGridColumn.cs#L1142, which would reduce reflection usage quite a bit.

Other than this, any refactorings of DataGrid control are not planned.

**Comment by dameng324** [comment]
@maxkatz6 Thanks for the info.

solving this by xaml compiler will be really nice.

**Comment by dameng324** [comment]
@maxkatz6 I wonder when will it be published? I didn't see this in the latest 11.2 rc version.

**Pull request merged**